backend.go

```diff
@@ -800,50 +808,110 @@ type DecodeData struct {
+	fragRangeLen := len(frags[0])
+	chunkSize := pieceSize + backend.headerSize
+	nrChunks := fragRangeLen / chunkSize
+	if nrChunks*chunkSize != fragRangeLen {
+		nrChunks++
 	}
 
-	// allocate output buffer
-	dataB, data := backend.pool.New(numBlock * piecesize * backend.K)
+	/* Fragments are sorted beforehand with the index of the first chunk.
+	   All chunks of a fragments share the same index. */
+	fragsIndex := make([]int, len(frags))
+	for i := 0; i < len(frags); i += 1 {
+		fragsIndex[i] = i
+	}
+
+	sort.Slice(fragsIndex, func(i, j int) bool {
+		lhs := frags[fragsIndex[i]]
+		rhs := frags[fragsIndex[j]]
 
+		var lhsIdx, rhsIdx C.int
+		lhsIdx = C.get_fragment_idx((*C.char)(unsafe.Pointer(&lhs[0])))
+		rhsIdx = C.get_fragment_idx((*C.char)(unsafe.Pointer(&rhs[0])))
+
+		return lhsIdx < rhsIdx
+	})
+
+	/* There is no reconstruction that can happen.
+	   All coding fragments must be ignored */
+	lastDataFragIdxExcl := 0
+	previousFragIdx := -1
+	for lastDataFragIdxExcl < len(fragsIndex) {
+		frag := frags[fragsIndex[lastDataFragIdxExcl]]
+		idx := int(C.get_fragment_idx((*C.char)(unsafe.Pointer(&frag[0]))))
+
+		if idx < 0 {
+			return nil, errors.New("invalid fragment index")
+		}
+
+		if idx >= backend.K {
+			break
+		}
+
+		if len(frags[0]) != fragRangeLen {
+			return nil, errors.New("invalid fragment len")
+		}
+
+		if previousFragIdx > 0 && (idx-previousFragIdx) != 1 {
+			/* Fragments are not contiguous. This functions doesn't supports
+			   gaps. */
+			return nil, errors.New("gaps in the provided fragments")
+		}
+
+		lastDataFragIdxExcl += 1
+		previousFragIdx = idx
+	}
+	fragsIndex = fragsIndex[:lastDataFragIdxExcl]
+
+	dataB, data := backend.pool.New(nrChunks * pieceSize * len(fragsIndex))
 	errorNb := uint32(0)
 	totLen := uint64(0)
-	wg.Add(numBlock)
+	wg.Add(nrChunks)
 
-	for i := 0; i < numBlock; i++ {
+	for i := 0; i < nrChunks; i++ {
 		// launch goroutines, providing them a subrange of the final buffer so it can be used
 		// in concurrency without need to lock it access
-		go func(blockNr int) {
-			cFrags := C.makeStrArray(C.int(len(frags)))
+		func(chunkIdx int) {
+			cDataFrags := C.makeStrArray(C.int(len(fragsIndex)))
 			// prepare the C array of pointer, respecting the offset in each fragments
-			for index, frags := range frags {
-				cSetArrayItem(cFrags, index, (*C.char)(unsafe.Pointer(&frags[blockNr*lenBlock])))
+
+			for i, idx := range fragsIndex {
+				frag := frags[idx]
+				cSetArrayItem(cDataFrags, i, (*C.char)(unsafe.Pointer(&frag[chunkIdx*chunkSize])))
 			}
 			// try to decode fastly (if we have all data fragments), providing the good offset of the
 			// linearized buffer, according the block number we are decoding
 			var outlen C.uint64_t
-			p := C.decode_fast(C.int(backend.K), cFrags, C.int(len(frags)),
-				(*C.char)(unsafe.Pointer(&data[blockNr*piecesize*backend.K])),
-				C.uint64_t(piecesize*backend.K), &outlen)
+			p := C.linearize(C.int(backend.K), cDataFrags, C.int(len(fragsIndex)),
```

Comment:
@benoit-a We talked about this but I don't understand why there is no need for `runtime.KeepAlive` for `frags` . It wasn't here before so I guess it is not necessary, but an explanation would be nice :)

Reply:
Added a `runtime.KeepAlive`. Previous implementation didn't have this issue as it had the fallback code in case of failure of which directly used the frags array. This is not the case here.